src/sys/macos.rs

```diff
@@ -0,0 +1,133 @@
+        Point {
+            x: other.x as _,
+            y: other.y as _,
+        }
+    }
+}
+
+impl Into<CGPoint> for Point {
+    fn into(self) -> CGPoint {
+        CGPoint::new(self.x as _, self.y as _)
+    }
+}
+
+#[derive(Debug)]
+pub enum Error<'a> {
+    CGDisplayMoveCursorToPoint(CGError),
+    CGEventNotCreated,
+    CGEventSourceStateInvalid,
+
+    InvalidButtonStr(&'a str),
+}
+
+impl<'a> error::Error for Error<'a> {}
+
+impl<'a> fmt::Display for Error<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::CGDisplayMoveCursorToPoint(err) => {
+                write!(f, "error in call to CGDisplayMoveCursorToPoint: {}", err)
+            }
+            Error::CGEventNotCreated => write!(f, "CGEvent could not be created"),
+            Error::CGEventSourceStateInvalid => write!(f, "invalid CGEventSourceStateID"),
+
+            Error::InvalidButtonStr(button) => write!(f, "invalid button str: {:?}", button),
+        }
+    }
+}
+
+pub struct Mouse;
+
+impl Mouse {
+    fn event_source<'a>() -> Result<CGEventSource, Error<'a>> {
+        Ok(
+            CGEventSource::new(CGEventSourceStateID::CombinedSessionState)
+                .or(Err(Error::CGEventSourceStateInvalid))?,
+        )
+    }
+
+    pub fn new() -> Mouse {
+        Mouse
+    }
+
+    pub fn move_to(&self, x: i32, y: i32) -> Result<(), Box<dyn error::Error>> {
+        let point = CGPoint::new(x as _, y as _);
+
+        match unsafe { CGDisplayMoveCursorToPoint(CGMainDisplayID(), point) } {
+            0 => Ok(()),
+            err => Err(Box::new(Error::CGDisplayMoveCursorToPoint(err))),
+        }
+    }
+
+    pub fn press<'a>(&self, button: &'a str) -> Result<(), Box<dyn error::Error + 'a>> {
+        let (event_type, mouse_button) = match button {
+            _ if button == LEFT => Ok((CGEventType::LeftMouseDown, CGMouseButton::Left)),
+            _ if button == MIDDLE => Ok((CGEventType::LeftMouseDown, CGMouseButton::Center)),
+            _ if button == RIGHT => Ok((CGEventType::RightMouseDown, CGMouseButton::Right)),
+            _ => Err(Box::new(Error::InvalidButtonStr(button))),
+        }?;
+
+        CGEvent::new_mouse_event(
+            Self::event_source()?,
+            event_type,
+            self.get_position()?.into(),
+            mouse_button,
+        )
+        .or(Err(Error::CGEventNotCreated))?
+        .post(CGEventTapLocation::HID);
+
+        Ok(())
+    }
+
+    pub fn release<'a>(&self, button: &'a str) -> Result<(), Box<dyn error::Error + 'a>> {
+        let (event_type, mouse_button) = match button {
+            _ if button == LEFT => Ok((CGEventType::LeftMouseUp, CGMouseButton::Left)),
+            _ if button == MIDDLE => Ok((CGEventType::LeftMouseUp, CGMouseButton::Center)),
+            _ if button == RIGHT => Ok((CGEventType::RightMouseUp, CGMouseButton::Right)),
+            _ => Err(Box::new(Error::InvalidButtonStr(button))),
+        }?;
+
```

Comment:
Is it possible to combine the duplicate from the latter comment into one method?

Reply:
I actually thought about that, but given that one is for up and the other for down variants the only thing I can imagine is more state (like passing a `bool` along to signify up/down) or some nasty `concat!` business. But hey, you got me to look up `CKMouseButton`, and turns out that it has a different `CGEventType` than the other two!